[Bug]: sessions.list cache ignores external session identity mutations

### Bug type

Behavior bug (incorrect output/state without crash)

### Beta release blocker

No

### Summary

`sessions.list` can return a completed cached result after an external session identity write, silently omitting the new or moved session.

### Steps to reproduce

1. On current `main` `2dd0e9b950462acc1f24fa9b207e9b7b0b4bd36b`, create a Gateway request context and call `sessions.list` to populate its completed-result cache.
2. Outside that Gateway context, call `upsertSessionEntry` with a new session key and identity.
3. Call `sessions.list` again with the same client and request.
4. Observe that the second call returns the same cached object and does not include the new session key.

### Expected behavior

Any committed session identity create, delete, move, replace, or reset should fence reuse of a completed `sessions.list` projection, including writes that occur outside the current Gateway request context.

### Actual behavior

The current cache fence tracks Gateway-context session mutations, live agent-run changes, and title-projection availability, but not the process-wide session identity mutation owner. The second list call returns the first cached object unchanged.

### OpenClaw version

Current `main` commit `2dd0e9b950462acc1f24fa9b207e9b7b0b4bd36b`

### Operating system

macOS

### Install method

Source checkout via `node scripts/run-vitest.mjs`

### Model

N/A — deterministic session/Gateway cache test with no model call

### Provider / routing chain

N/A — in-process Gateway method and SQLite session accessor

### Additional provider/model setup details

No external provider or credential is used.

### Logs

```shell
sessions-read-cache.test.ts (21 tests | 1 failed | 20 skipped)
× invalidates a completed result after an external session identity mutation
AssertionError: expected refreshed not to be first // Object.is equality
Test Files  1 failed (1)
Tests       1 failed | 20 skipped (21)
```

### Screenshots, recordings, and evidence

The exact regression passes when the cache fence includes the monotonic process-wide session identity mutation version. The complete session-list cache, lifecycle-event, and accessor suites pass 117/117, and 20 fresh-process regression runs pass 20/20.

### Impact and severity

Affected: Gateway clients and internal consumers that call `sessions.list` after a plugin, cron job, session accessor, or other external owner creates or changes session identity.

Severity: Medium; the API can silently omit a valid session.

Frequency: Deterministic while the completed cache remains reusable and no separately fenced mutation occurs.

Consequence: session pickers, automation, and operators can act on stale session inventory.

### Additional information

The process already emits authoritative identity mutations from the SQLite session-accessor owner. A monotonic version at that event boundary lets projections fence external writes without coupling every writer to a specific Gateway request context.

--- a/src/gateway/server-methods/sessions-list-cache.ts
+++ b/src/gateway/server-methods/sessions-list-cache.ts
@@ -1,6 +1,7 @@
 import type { SessionsListParams } from "../../../packages/gateway-protocol/src/index.js";
 import type { OpenClawConfig } from "../../config/types.openclaw.js";
 import { readAgentRunIndexVersion } from "../../infra/agent-run-registry.js";
+import { readSessionIdentityMutationVersion } from "../../sessions/session-lifecycle-events.js";
 import { isGatewayAdmin } from "../session-sharing.js";
 import { readSessionTitleProjectionUnavailableVersion } from "../session-transcript-title-reader.js";
 import type { SessionsListResult } from "../session-utils.types.js";
@@ -10,6 +11,7 @@ import type { GatewayClient, GatewayRequestContext, RespondFn } from "./types.js
 
 type SessionListFence = {
   agentRunIndexVersion: number;
+  sessionIdentityMutationVersion: number;
   sessionsMutationVersion: number;
   titleProjectionUnavailableVersion: number;
 };
@@ -27,6 +29,7 @@ const sessionListsByContext = new WeakMap<GatewayRequestContext, SessionListStat
 function readSessionListFence(context: GatewayRequestContext): SessionListFence {
   return {
     agentRunIndexVersion: readAgentRunIndexVersion(),
+    sessionIdentityMutationVersion: readSessionIdentityMutationVersion(),
     sessionsMutationVersion: readSessionsMutationVersion(context),
     titleProjectionUnavailableVersion: readSessionTitleProjectionUnavailableVersion(),
   };
@@ -35,6 +38,7 @@ function readSessionListFence(context: GatewayRequestContext): SessionListFence 
 function matchesSessionListFence(value: SessionListFence, fence: SessionListFence): boolean {
   return (
     value.agentRunIndexVersion === fence.agentRunIndexVersion &&
+    value.sessionIdentityMutationVersion === fence.sessionIdentityMutationVersion &&
     value.sessionsMutationVersion === fence.sessionsMutationVersion &&
     value.titleProjectionUnavailableVersion === fence.titleProjectionUnavailableVersion
   );
@@ -112,8 +116,8 @@ export async function respondWithCachedSessionList(params: {
 }): Promise<void> {
   const workKey = sessionListWorkKey(params.request, params.client);
   const state = sessionListState(params.context, params.config);
-  // Every input that can change a projected row must fence reuse. Store mutations and
-  // live-run transitions have separate owners, so their monotonic counters stay separate.
+  // Every input that can change a projected row must fence reuse. Session identity,
+  // Gateway projection, and live-run mutations have separate monotonic owners.
   const fence = readSessionListFence(params.context);
   // Activity windows and child retention expire without mutations; hidden paginated rows
   // prevent deriving a safe deadline, so only concurrent temporal requests share work.

--- a/src/gateway/server-methods/sessions-read-cache.test.ts
+++ b/src/gateway/server-methods/sessions-read-cache.test.ts
@@ -290,6 +290,31 @@ describe("sessions.list single-flight", () => {
     });
   });
 
+  it("invalidates a completed result after an external session identity mutation", async () => {
+    await withOpenClawTestState({ scenario: "minimal" }, async () => {
+      const config = await seedSessions();
+      const context = requestContext(config);
+      const client = identifiedClient("owner@example.com");
+      const request = { archived: "all" as const, limit: 100 };
+
+      const first = await listSessions({ client, context, request });
+      await upsertSessionEntry(
+        { agentId: "main", sessionKey: "agent:main:external" },
+        {
+          sessionId: "main-external",
+          updatedAt: 500,
+          createdActor: { type: "human", id: "owner@example.com" },
+          visibility: "shared",
+        },
+      );
+      const refreshed = await listSessions({ client, context, request });
+
+      expect(refreshed).not.toBe(first);
+      expect(loader.calls).toHaveBeenCalledTimes(2);
+      expect(refreshed.sessions.map((session) => session.key)).toContain("agent:main:external");
+    });
+  });
+
   it("expires completed rows at the earliest projected agent-status deadline", async () => {
     await withOpenClawTestState({ scenario: "minimal" }, async () => {
       const clock = vi.spyOn(Date, "now").mockReturnValue(1_000);

--- a/src/sessions/session-lifecycle-events.test.ts
+++ b/src/sessions/session-lifecycle-events.test.ts
@@ -1,6 +1,12 @@
 // Session lifecycle event tests cover lifecycle event ordering and serialization.
 import { describe, expect, it } from "vitest";
-import { emitSessionLifecycleEvent, onSessionLifecycleEvent } from "./session-lifecycle-events.js";
+import {
+  emitSessionIdentityMutation,
+  emitSessionLifecycleEvent,
+  onSessionIdentityMutation,
+  onSessionLifecycleEvent,
+  readSessionIdentityMutationVersion,
+} from "./session-lifecycle-events.js";
 
 function createListenerSpy(options: { throws?: boolean } = {}) {
   const calls: unknown[][] = [];
@@ -16,6 +22,27 @@ function createListenerSpy(options: { throws?: boolean } = {}) {
 }
 
 describe("session lifecycle events", () => {
+  it("advances the identity mutation version before notifying listeners", () => {
+    const previousVersion = readSessionIdentityMutationVersion();
+    const observedVersions: number[] = [];
+    const unsubscribe = onSessionIdentityMutation(() => {
+      observedVersions.push(readSessionIdentityMutationVersion());
+    });
+
+    try {
+      emitSessionIdentityMutation({
+        kind: "create",
+        previous: { sessionKeys: [] },
+        current: { sessionId: "session-1", sessionKeys: ["agent:main:external"] },
+      });
+    } finally {
+      unsubscribe();
+    }
+
+    expect(readSessionIdentityMutationVersion()).toBe(previousVersion + 1);
+    expect(observedVersions).toEqual([previousVersion + 1]);
+  });
+
   it("delivers events to active listeners and stops after unsubscribe", () => {
     const { calls, listener } = createListenerSpy();
     const unsubscribe = onSessionLifecycleEvent(listener);

--- a/src/sessions/session-lifecycle-events.ts
+++ b/src/sessions/session-lifecycle-events.ts
@@ -1,5 +1,5 @@
 /** Session lifecycle event broadcast to observers when a session is created or linked. */
-import { resolveGlobalSet } from "../shared/global-singleton.js";
+import { resolveGlobalSet, resolveGlobalSingleton } from "../shared/global-singleton.js";
 export type SessionLifecycleEvent = {
   sessionKey: string;
   reason: string;
@@ -36,6 +36,10 @@ const SESSION_IDENTITY_MUTATION_LISTENERS = resolveGlobalSet<SessionIdentityMuta
   Symbol.for("openclaw.sessionIdentityMutationListeners"),
   "close-and-restart",
 );
+const SESSION_IDENTITY_MUTATION_STATE = resolveGlobalSingleton(
+  Symbol.for("openclaw.sessionIdentityMutationState"),
+  () => ({ version: 0 }),
+);
 
 /** Registers a session lifecycle listener. */
 export function onSessionLifecycleEvent(listener: SessionLifecycleListener): () => void {
@@ -63,7 +67,13 @@ export function onSessionIdentityMutation(listener: SessionIdentityMutationListe
   };
 }
 
+/** Monotonic fence for projections that consume session identities across owner boundaries. */
+export function readSessionIdentityMutationVersion(): number {
+  return SESSION_IDENTITY_MUTATION_STATE.version;
+}
+
 export function emitSessionIdentityMutation(mutation: SessionIdentityMutation): void {
+  SESSION_IDENTITY_MUTATION_STATE.version += 1;
   for (const listener of SESSION_IDENTITY_MUTATION_LISTENERS) {
     try {
       listener(mutation);

